File: server/app/routes/user_routes.py
from flask import Blueprint, request, jsonify
from firebase_admin import auth, firestore
from app.config.firebaseSetup import get_firestore_client
from app.models.user_model import UserModel

user_bp = Blueprint("user", __name__, url_prefix="/user")
db = get_firestore_client()

@user_bp.route("/<user_id>", methods=["GET"])
def get_user(user_id):
    """
    Retrieve user details from Firestore by user_id.
    """
    user_ref = db.collection("users").document(user_id).get()
    
    if not user_ref.exists:
        return jsonify({"error": "User not found"}), 404
    
    user_data = user_ref.to_dict()
    return jsonify(user_data), 200

@user_bp.route("/update/<user_id>", methods=["PUT"])
def update_user(user_id):
    """
    Update user details in Firestore.
    """
    user_ref = db.collection("users").document(user_id)
    
    if not user_ref.get().exists:
        return jsonify({"error": "User not found"}), 404

    updated_data = request.get_json()
    user_ref.update(updated_data)
    return jsonify({"message": "User updated successfully"}), 200

@user_bp.route("/delete/<user_id>", methods=["DELETE"])
def delete_user(user_id):
    """
    Delete user from Firestore and Firebase Authentication.
    """
    try:
        # Delete from Firebase Auth
        auth.delete_user(user_id)
        
        # Delete from Firestore
        db.collection("users").document(user_id).delete()
        
        return jsonify({"message": "User deleted successfully"}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 400

File: server/routes/userRoutes.js
import express from 'express';
import { getFirestore } from 'firebase-admin/firestore';
import { getStorage } from 'firebase-admin/storage';
import multer from 'multer';

const storage = getStorage();
const router = express.Router();
const db = getFirestore();
const upload = multer({ storage: multer.memoryStorage() });

// ✅ Get User Profile (Now Includes Plants)
router.get('/profile/:userId', async (req, res) => {
    try {
        const userRef = db.collection('users').doc(req.params.userId);
        const userDoc = await userRef.get();

        if (!userDoc.exists) return res.status(404).json({ error: 'User profile not found' });

        // Fetch user plants
        const plantsSnapshot = await userRef.collection('plants').get();
        const plants = plantsSnapshot.docs.map(doc => ({ id: doc.id, ...doc.data() }));

        res.json({ ...userDoc.data(), plants });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// ✅ Add a New Plant to User Profile
router.post('/profile/:userId/plants', async (req, res) => {
    try {
        const { userId } = req.params;
        const { name, type, wateringSchedule, lastWatered, healthStatus } = req.body;
        
        if (!name || !type || !wateringSchedule || !lastWatered) {
            return res.status(400).json({ error: 'Missing required fields' });
        }

        const plantRef = db.collection('users').doc(userId).collection('plants').doc();
        await plantRef.set({
            name,
            type,
            wateringSchedule,
            lastWatered,
            healthStatus: healthStatus || 'Healthy',
            createdAt: new Date().toISOString()
        });

        res.status(201).json({ success: true, id: plantRef.id });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// ✅ Get Chat History
router.get('/chat-history/:userId', async (req, res) => {
    try {
        const messages = [];
        const messageRefs = await db.collection(`chats/${req.params.userId}/messages`)
            .orderBy('timestamp')
            .get();

        messageRefs.forEach((msg) => messages.push(msg.data()));

        res.json(messages);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// ✅ Save Chat Message
router.post('/message', async (req, res) => {
    try {
        const { userId, message } = req.body;
        if (!userId || !message) return res.status(400).json({ error: 'Missing required fields' });

        await db.collection(`chats/${userId}/messages`).add({
            ...message,
            timestamp: new Date()
        });

        res.json({ success: true });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

router.post('/profile/:userId/avatar', upload.single('avatar'), async (req, res) => {
    console.log("✅ Hit /profile/:userId/avatar route");

    try {
      const file = req.file;
      const { userId } = req.params;
  
      if (!file) return res.status(400).json({ error: 'No file uploaded' });
  
      const bucket = storage.bucket();
      const fileName = `avatars/${userId}/${Date.now()}_${file.originalname}`;
      const fileUpload = bucket.file(fileName);
  
      const stream = fileUpload.createWriteStream({
        metadata: {
          contentType: file.mimetype,
        },
      });
  
      stream.on('error', (err) => {
        return res.status(500).json({ error: err.message });
      });
  
      stream.on('finish', async () => {
        await fileUpload.makePublic();
        const publicUrl = `https://storage.googleapis.com/${bucket.name}/${fileName}`;
  
        await db.collection('users').doc(userId).update({
          profileImage: publicUrl,
          updatedAt: new Date().toISOString()
        });
  
        res.status(200).json({ success: true, imageUrl: publicUrl });
      });
  
      stream.end(file.buffer);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  });

export default router;
